--- a/background_worker_build.py	2024-05-16 19:27:13.000000000 +0800
+++ b/background_worker_build.py	2024-05-16 19:27:13.000000000 +0800
@@ -27,7 +27,7 @@
     call_copr_repo, pkg_name_evr, run_cmd, register_build_result,
 )
 from copr_backend.job import BuildJob
-from copr_backend.msgbus import MessageSender
+from copr_backend.euler_msgbus import MessageSender
 from copr_backend.sign import sign_rpms_in_dir, get_pubkey
 from copr_backend.sshcmd import SSHConnection, SSHConnectionError
 from copr_backend.vm_alloc import ResallocHostFactory
