diff --git a/backend/copr_backend/helpers.py b/backend/copr_backend/helpers.py
index df2819e0..c85c7396 100644
--- a/backend/copr_backend/helpers.py
+++ b/backend/copr_backend/helpers.py
@@ -288,6 +288,9 @@ class BackendConfigReader(object):
         opts.redis_port = _get_conf(
             cp, "backend", "redis_port", "6379")
 
+        opts.redis_pwd = _get_conf(
+            cp, "backend", "redis_password", None)
+
         opts.redis_db = _get_conf(
             cp, "backend", "redis_db", "0")

